client/src/components/annotators/ImageAnnotator.vue

```diff
@@ -97,37 +114,104 @@ export default {
           this.syncedFrame = this.maxFrame;
           return;
         }
+        if (!this.imgs[this.frame].cached || this.loadingVideo) {
+          this.frame -= 1;
+          this.loadingVideo = true;
+          return;
+        }
         this.seek(this.frame);
         setTimeout(this.syncWithVideo, 1000 / this.frameRate);
       }
     },
-    cacheImage() {
+    async cacheImage() {
       const { frame } = this;
-      const max = Math.min(frame + 10, this.maxFrame);
       const { imgs } = this;
+      const cachedFrames = this.cacheSeconds * this.frameRate;
+      const min = Math.floor(Math.max(0, frame - cachedFrames * (1 - this.frontBackRatio)));
+      const max = Math.ceil(Math.min(frame + this.frontBackRatio * cachedFrames, this.maxFrame));
+      const frameDiff = Math.abs(this.frame - this.lastFrame);
+      const prevFrame = (this.frame < this.lastFrame);
       this.pendingImgs.forEach((imageAndFrame) => {
-        if (imageAndFrame[1] < frame || imageAndFrame[1] > max) {
+        // the current loading cache nees to be wiped out if we seek forward, backwards or
+        // if we are out of the current range of the cache
+        if (imageAndFrame[1] < min || imageAndFrame[1] > max || frameDiff > 2 || prevFrame) {
           imgs[imageAndFrame[1]] = null;
           // eslint-disable-next-line no-param-reassign
           imageAndFrame[0].src = '';
           this.pendingImgs.delete(imageAndFrame);
         }
       });
-      for (let i = frame; i <= max; i += 1) {
-        if (!imgs[i]) {
+      // if not playing we want the seeked to frame immediately loaded for UX
+      if (!this.playing && !imgs[frame] && frame > 0) {
+        await this.loadFrame(frame);
+      }
+      // Cache a new range based on the parameters
+      this.cacheNewRange(min, max);
+    },
+    /**
+     * Wraps loading of a single frame in a promise
+     * @param {int} frame number to be loaded
+     * @returns {Promise} resolves when the image from the frame is loaded
+     */
+    loadFrame(frame) {
+      return new Promise((resolve) => {
+        const img = new Image();
+        img.crossOrigin = 'Anonymous';
+        img.src = this.imageUrls[frame];
+        this.imgs[frame] = img;
+        img.onload = () => {
+          img.onload = null;
+          img.cached = true;
+          resolve(frame);
+        };
+      });
+    },
+    /**
+     * Caches a new range of frames to load
+     * @param {int} min - lower bound frame number for caching
+     * @param {int} max - upper bound frame number for caching
+     */
+    cacheNewRange(min, max) {
```

Comment:
It looks like the cache loads back to front, starting at currentFrame - (0.1*cachesize).  Would it make a difference to load from the current frame outward sort of jumping back and fourth?

Reply:
that's why it's a draft, one of my ideas was loading from the current frame out.  I haven't implemented that yet.